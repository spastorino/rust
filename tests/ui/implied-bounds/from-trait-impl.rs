// Foo<Vec<X>> shouldn't imply X: 'static.
// We don't use region constraints from trait impls in implied bounds.

trait Trait {
    type Assoc;
}

impl<X: 'static> Trait for Vec<X> {
    type Assoc = ();
}

struct Foo<T: Trait>(T)
where
    T::Assoc: 'static, // any predicate naming T::Assoc
;

fn foo<X>(_: Foo<Vec<X>>) {}
//~^ ERROR `X` may not live long enough
//~| ERROR `X` may not live long enough

fn main() {}
