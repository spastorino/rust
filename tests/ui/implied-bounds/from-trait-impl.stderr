error[E0310]: the parameter type `X` may not live long enough
  --> $DIR/from-trait-impl.rs:17:1
   |
LL | fn foo<X>(_: Foo<Vec<X>>) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^
   | |
   | the parameter type `X` must be valid for the static lifetime...
   | ...so that the type `X` will meet its required lifetime bounds
   |
help: consider adding an explicit lifetime bound
   |
LL | fn foo<X: 'static>(_: Foo<Vec<X>>) {}
   |         +++++++++

error[E0310]: the parameter type `X` may not live long enough
  --> $DIR/from-trait-impl.rs:17:14
   |
LL | fn foo<X>(_: Foo<Vec<X>>) {}
   |              ^^^^^^^^^^^
   |              |
   |              the parameter type `X` must be valid for the static lifetime...
   |              ...so that the type `X` will meet its required lifetime bounds
   |
help: consider adding an explicit lifetime bound
   |
LL | fn foo<X: 'static>(_: Foo<Vec<X>>) {}
   |         +++++++++

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0310`.
